src/sec.rs

```diff
@@ -993,7 +997,55 @@ impl Sec {
     fn saga_finished(&mut self, done_data: SagaDoneData) {
         let saga_id = done_data.saga_id;
         let saga = self.sagas.remove(&saga_id).unwrap();
-        info!(&saga.log, "saga finished");
+
+        // Log every saga completion, regardless of what happened:
+        //
+        // * There will be exactly one log entry for every saga that finishes.
+        // * The message will always say "saga finished".
+        // * There is always a "result" field that's either "success" or
+        //   "failure".
+        //
+        // There are three cases:
+        //
+        // 1. Successful completion (no errors).  This will be an "info" level
+        //    message.
+        //
+        // 2. Clean failure (no errors during unwinding).  This will be a "warn"
+        //    level message and have "undo_result" = "success".
+        //
+        // 3. Stuck (an error was generated during unwinding).  This will be an
+        //    "error" level message and have "undo_result" = "failure".
+        //
+        // This should make it fairly straightforward to summarize the results
+        // programmatically.  The saga id and name are already in the saga.log
+        // object so that will also make it easy to summarize by saga name or to
+        // find the entry for a particular saga.
+        match &done_data.result.kind {
+            Ok(SagaResultOk { .. }) => {
+                info!(&saga.log, "saga finished"; "result" => "success");
+            }
+            Err(SagaResultErr {
+                error_node_name,
+                error_source,
+                undo_failure,
+            }) => {
+                let log = saga.log.new(o!(
+                    "result" => "failure",
+                    "action_error_node_name" =>
+                        format!("{:?}", error_node_name),
+                    "action_error_source" => format!("{:?}", error_source),
+                ));
```

Comment:
a little bit unconvinced about constructing a child `Logger` with these fields and then immediately logging one record with it, rather than just repeating them in both log macros. the child logger will (as I understand it) require at least three allocations (one for the `BTreeMap` and one for each `format!` macro), while just including them in the `error!` macros needn't allocate at all. 

i know this code, and especially the "saga failed" code path, may not be particularly performance sensitive, so i'm fine with leaving this as-is, but i just figured i'd mention it.

Reply:
This is a fair point.  Here I prefer to prioritize presence, consistency, and accuracy of the information in this log message over straight-line performance to that degree.